extensions/ql-vscode/src/query-history.ts

```diff
@@ -162,120 +168,284 @@ export class QueryHistoryManager {
+      this.treeDataProvider.refresh();
+    });
+
+    // displays query text in a read-only document
+    vscode.workspace.registerTextDocumentContentProvider('codeql', {
+      provideTextDocumentContent(
+        uri: vscode.Uri
+      ): vscode.ProviderResult<string> {
+        const params = new URLSearchParams(uri.query);
+
+        return (
+          (JSON.parse(params.get('isQuickEval') || '')
+            ? SHOW_QUERY_TEXT_QUICK_EVAL_MSG
+            : SHOW_QUERY_TEXT_MSG) + params.get('queryText')
+        );
+      },
+    });
+  }
+
   async invokeCallbackOn(queryHistoryItem: CompletedQuery) {
     if (this.selectedCallback !== undefined) {
       const sc = this.selectedCallback;
       await sc(queryHistoryItem);
     }
   }
 
-  async handleOpenQuery(queryHistoryItem: CompletedQuery): Promise<void> {
-    const textDocument = await vscode.workspace.openTextDocument(vscode.Uri.file(queryHistoryItem.query.program.queryPath));
-    const editor = await vscode.window.showTextDocument(textDocument, vscode.ViewColumn.One);
-    const queryText = queryHistoryItem.options.queryText;
-    if (queryText !== undefined && queryHistoryItem.options.isQuickQuery) {
-      await editor.edit(edit => edit.replace(textDocument.validateRange(
-        new vscode.Range(0, 0, textDocument.lineCount, 0)), queryText)
+  async handleOpenQuery(
+    singleItem: CompletedQuery,
+    multiSelect: CompletedQuery[]
+  ): Promise<void> {
+    if (!this.assertSingleQuery(multiSelect)) {
+      return;
+    }
+
+    const textDocument = await vscode.workspace.openTextDocument(
+      vscode.Uri.file(singleItem.query.program.queryPath)
+    );
+    const editor = await vscode.window.showTextDocument(
+      textDocument,
+      vscode.ViewColumn.One
+    );
+    const queryText = singleItem.options.queryText;
+    if (queryText !== undefined && singleItem.options.isQuickQuery) {
+      await editor.edit((edit) =>
+        edit.replace(
+          textDocument.validateRange(
+            new vscode.Range(0, 0, textDocument.lineCount, 0)
+          ),
+          queryText
+        )
       );
     }
   }
 
-  async handleRemoveHistoryItem(queryHistoryItem: CompletedQuery) {
-    this.treeDataProvider.remove(queryHistoryItem);
-    queryHistoryItem.dispose();
+  async handleRemoveHistoryItem(
+    singleItem: CompletedQuery,
+    multiSelect: CompletedQuery[]
+  ) {
+    (multiSelect || [singleItem]).forEach((item) => {
+      this.treeDataProvider.remove(item);
+      item.dispose();
+    });
     const current = this.treeDataProvider.getCurrent();
     if (current !== undefined) {
       this.treeView.reveal(current);
       await this.invokeCallbackOn(current);
     }
   }
 
-  async handleSetLabel(queryHistoryItem: CompletedQuery) {
+  async handleSetLabel(
+    singleItem: CompletedQuery,
+    multiSelect: CompletedQuery[]
+  ): Promise<void> {
+    if (!this.assertSingleQuery(multiSelect)) {
```

Comment:
Since labels can be `sprintf`-like patterns, I think we should support setting group of query history items' label at once.

Reply:
I feel like that's a bit confusing. If someone creates a label without a sprintf pattern, all of the labels will be the same and you can no longer distinguish between them.

I guess, I _could_ special case things where if you multi-select a label change, the second one is suffixed with `-2` and then `-3`, etc.

Reply:
We discussed this out of band and I backed off from thinking it's important to have this. I'll make a PR if I think I have a good solution.